File: packages/whisker-asset/ios/Sources/WhiskerAsset/AssetModule.swift
// `whisker-asset` Module (iOS). View-less, startup-only.
//
// Installs the runtime asset-resolution base so `asset!("images/x.png")`
// — which lowers to `whisker_asset::resolve("images/x.png")` in Rust —
// composes a directly-loadable `file://` URL into the app bundle.
//
// ## Why native, and why here
//
// The iOS base is the app bundle's `whisker_assets/` folder reference,
// placed there by the whisker-asset build plugin. Its absolute path is
// `Bundle.main.bundlePath`, which is only known at RUNTIME — so unlike
// Android, where the Rust side installs a fixed `file:///android_asset/…`
// constant itself, iOS must hand the path to Rust from native at launch.
//
// The value is a `file://` URL STRING, not a bare filesystem path, so
// what Kingfisher receives in `WhiskerImageView.setSrc` is directly
// loadable as a `URL`.
//
// ## Startup hook
//
// The DSL has no explicit `OnCreate`, but the framework reads each
// `Module` subclass's `definitionLazy` exactly once at registration time
// — app launch, before the first render. Installing the base inside
// `definition()` therefore runs ahead of any `resolve` call and is
// memoised against re-running.

import Foundation
import WhiskerModule

/// Rust C-ABI export, defined in `packages/whisker-asset/src/lib.rs`:
///
/// ```c
/// void whisker_asset_set_ios_base(const uint8_t *ptr, size_t len);
/// ```
///
/// Exported from the app's `cdylib`, since the `#[whisker::main]` crate
/// transitively links `whisker-asset` via `asset!(…)`. `@_silgen_name`
/// binds the raw C symbol by name, resolved by the dynamic linker at
/// load, without needing it in a C header / module map. `ptr`/`len`
/// describe a non-NUL-terminated UTF-8 buffer; Rust copies the bytes, so
/// a borrow for the duration of the call is enough.
@_silgen_name("whisker_asset_set_ios_base")
private func whisker_asset_set_ios_base(_ ptr: UnsafePointer<UInt8>?, _ len: Int)

@WhiskerModule
public final class AssetModule: Module {

    public override func definition() -> ModuleDefinition {
        Self.installBase()

        return ModuleDefinition {
            Name("WhiskerAsset")
        }
    }

    /// Compose `file://<bundlePath>/whisker_assets` and hand it to the Rust
    /// resolver via the C export.
    ///
    /// The `whisker_assets` segment must match the namespace the build
    /// plugin bundles folder-referenced assets under, so that `IosDir`'s
    /// `"{dir}/{rel}"` composition yields
    /// `file://<bundle>/whisker_assets/<rel>`.
    static func installBase() {
        let base = "file://\(Bundle.main.bundlePath)/whisker_assets"
        // Copied into an Array for a stable pointer across the call.
        let bytes = Array(base.utf8)
        bytes.withUnsafeBufferPointer { buf in
            whisker_asset_set_ios_base(buf.baseAddress, buf.count)
        }
    }
}

File: packages/whisker-asset/android/src/main/kotlin/rs/whisker/modules/asset/AssetModule.kt
// `whisker-asset` Module (Android). View-less, declaration-only.
//
// ## Where the asset base actually gets installed
//
// The Android resolver base is the fixed constant
// `file:///android_asset/whisker/` — there is no runtime data to ferry
// across the Kotlin↔Rust boundary (contrast iOS, whose bundle path is
// only known at runtime and so MUST come from native Swift). Installing
// it therefore happens entirely on the Rust side: `whisker-asset`'s lib
// registers an `.init_array` constructor (`install_android_base`) that
// runs the instant the app's `.so` is loaded via
// `System.loadLibrary` — well before the first render or any
// `asset!(…)` → `whisker_asset::resolve(…)` call. See
// `packages/whisker-asset/src/lib.rs`.
//
// Calling the `whisker_asset_set_android()` C export from Kotlin would
// require a JNI C++ shim (Kotlin `external fun` resolves only
// `Java_*`-named symbols, not bare C symbols), and that shim lives in
// the driver bridge crate, not here — so the Rust constructor is the
// cleaner, self-contained path. The C export remains available for a
// future host that wants to flip the base explicitly.
//
// This Kotlin module exists so the `[package.metadata.whisker]` marker
// wires a real Android subproject (KSP registration, parity with the
// iOS `AssetModule`) and as the home for any future native asset work.
// Its `definition()` only declares the module name.

package rs.whisker.modules.asset

import rs.whisker.runtime.Module
import rs.whisker.runtime.ModuleDefinition

public class AssetModule : Module() {
    public override fun definition(): ModuleDefinition = ModuleDefinition {
        Name("WhiskerAsset")
    }
}
